### PR TITLE
Axis Labels with offset Spines

### PR DESCRIPTION
**Pull request by cimarronm**
An attempt to fix the issue #3715. This takes into account the location of the axis spines when locating the axis label if there are no tick labels present.

Examples in #3715
###  Before patch
image

### After patch
image

### CODE REVIEW

**Comment by cimarronm** [comment]
Should probably add a test for this as well

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1743,7 +1745,9 @@ def _update_label_position(self, bboxes, bboxes2):
 
         else:
             if not len(bboxes2):
-                top = self.axes.bbox.ymax
+                top = (self.axes.spines['top'].get_transform().
```
Comment (jkseppan) [inline lib/matplotlib/axis.py]: The docstring of the function only talks about tick labels, now the function uses spines too

Comment (cimarronm) [inline lib/matplotlib/axis.py]: Good catch. Updated the docstrings

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1726,14 +1727,17 @@ def set_label_position(self, position):
...
-        boxes of all the ticklabels
+        boxes of all the ticklabels or using the axis spines if no
+        ticklabels are present
         """
         if not self._autolabelpos:
             return
         x, y = self.label.get_position()
         if self.label_position == 'bottom':
             if not len(bboxes):
-                bottom = self.axes.bbox.ymin
```
Comment (tacaswell) [inline lib/matplotlib/axis.py]: Would it make more sense to just throw the bounding box of the spine and the axes in to the bboxes list and let the `union` call take care of this logic?

Comment (cimarronm) [inline lib/matplotlib/axis.py]: Yeah, @tacaswell, I think that makes more sense. I'll update to do just that.

File: lib/matplotlib/axis.py [hunk 3]
```diff
@@ -1725,28 +1725,30 @@ def set_label_position(self, position):
...
         if not self._autolabelpos:
             return
         x, y = self.label.get_position()
         if self.label_position == 'bottom':
-            if not len(bboxes):
-                bottom = self.axes.bbox.ymin
-            else:
-                bbox = mtransforms.Bbox.union(bboxes)
-                bottom = bbox.y0
+            spinebbox = (self.axes.spines['bottom'].get_transform().
```
Comment (efiring) [inline lib/matplotlib/axis.py]: Style alternative: I would factor out `spine = self.axes.spines['bottom']` so as to make the spinebbox expression more compact.

**Comment by efiring** [comment]
Looks good.  A test would be nice.

File: lib/matplotlib/axis.py (hunk 3 above)
Comment (cimarronm) [inline lib/matplotlib/axis.py]: Yeah, that is much better. Will modify it as so

**Comment by cimarronm** [comment]
Working on a test

**Comment by tacaswell** [comment]
Thank you

**Pull request merged**